src/services/navigationBar.ts

```diff
@@ -2,860 +2,660 @@
 
 /* @internal */
 namespace ts.NavigationBar {
-    export function getNavigationBarItems(sourceFile: SourceFile, compilerOptions: CompilerOptions): ts.NavigationBarItem[]  {
-        // TODO: Handle JS files differently in 'navbar' calls for now, but ideally we should unify
-        // the 'navbar' and 'navto' logic for TypeScript and JavaScript.
-        if (isSourceFileJavaScript(sourceFile)) {
-            return getJsNavigationBarItems(sourceFile, compilerOptions);
-        }
-
-        return getItemsWorker(getTopLevelNodes(sourceFile), createTopLevelItem);
-
-        function getIndent(node: Node): number {
-            let indent = 1; // Global node is the only one with indent 0.
-
-            let current = node.parent;
-            while (current) {
-                switch (current.kind) {
-                    case SyntaxKind.ModuleDeclaration:
-                        // If we have a module declared as A.B.C, it is more "intuitive"
-                        // to say it only has a single layer of depth
-                        do {
-                            current = current.parent;
-                        }
-                        while (current.kind === SyntaxKind.ModuleDeclaration);
-
-                        // fall through
-                    case SyntaxKind.ClassDeclaration:
-                    case SyntaxKind.EnumDeclaration:
-                    case SyntaxKind.InterfaceDeclaration:
-                    case SyntaxKind.FunctionDeclaration:
-                        indent++;
-                }
-
-                current = current.parent;
-            }
+    /**
+     * Represents a navigation bar item and its children.
+     * The returned NavigationBarItem is more complicated and doesn't include 'parent', so we use these to do work before converting.
+     */
+    interface NavigationBarNode {
+        node: Node;
+        additionalNodes: Node[] | undefined;
+        parent: NavigationBarNode | undefined; // Present for all but root node
+        children: NavigationBarNode[] | undefined;
+        indent: number; // # of parents
+    }
 
-            return indent;
-        }
+    export function getNavigationBarItems(sourceFile_: SourceFile): NavigationBarItem[] {

```

Comment:
This may be pedantic, but I don't like the variable name `sourceFile_` - we generally don't use underscores. I'd prefer a variable name which describes how the SourceFile passed is used, or why it is required.